### ISSUE TITLE
ALX: a function that returns a pointer to a newly allocated space in memory, which contains a copy of the string given as a parameter (_strdup)

### ISSUE THREAD

**Issue by ahmed149** [ahmed149/learning-programing-in-C#55]
#include "main.h"
#include <stdlib.h>

/**
  * _strdup - Duplicate a string
  * @str: the string to duplicate
  *
  * Return: the string duplicated
  */
char *_strdup(char *str)
{
	int a = 0, i = 1;
	char *s;

	if (str == NULL)
		return (NULL);

	while (str[i])
	{
		i++;
	}

	s = malloc((sizeof(char) * i) + 1);

	if (s == NULL)
		return (NULL);

	while (a < i)
	{
		s[a] = str[a];
		a++;
	}

	s[a] = '\0';
	return (s);
}

**Comment by ahmed149** [comment]
This code defines a C function called `_strdup` that duplicates a string (a sequence of characters) passed as an argument. Here's a step-by-step explanation of the code:

1. `#include "main.h"` and `#include <stdlib.h>`: These lines include necessary header files. `"main.h"` is likely a custom header file that may contain additional declarations or definitions used in the program. `<stdlib.h>` is included for memory allocation functions.

2. The code includes a function documentation comment using a multi-line comment block. It describes the purpose of the function, its parameters, and its return value.

3. Function Signature:
   ```c
   char *_strdup(char *str)
   ```
   - `char *str`: This parameter represents the string that you want to duplicate.

4. Inside the function:
   - It declares two integer variables `a` and `i` and a character pointer `s`.
   - `a` is used as a loop counter.
   - `i` is initialized to 1 and is used to determine the length of the input string `str`.

5. The code checks if the input string `str` is `NULL`. If `str` is `NULL`, it means that no valid string is provided to duplicate, so it returns `NULL` to indicate an error.

6. It enters a `while` loop that iterates through the characters of the input string `str` to determine its length:
   - In each iteration, it increments `i` to count the number of characters until it reaches the null terminator `'\0'`, which signifies the end of the string.

7. After determining the length of the input string, it uses `malloc` to dynamically allocate memory for a new string (`s`) of the same length plus one additional character. The extra character is for the null terminator `'\0'` that terminates C strings.

8. It checks if the memory allocation using `malloc` was successful by testing if `s` is `NULL`. If memory allocation fails, it returns `NULL` to indicate an error.

9. It enters another `while` loop to copy the characters from the input string `str` to the newly allocated string `s`:
   - In each iteration, it copies the character from `str` to the corresponding position in `s`.
   - The loop continues until `a` is less than `i`, effectively copying all characters from `str`.

10. After copying all characters, it explicitly adds a null terminator `'\0'` to the end of the duplicated string `s` to ensure it's properly terminated as a C string.

11. Finally, it returns a pointer to the duplicated string `s`, which now contains a copy of the input string `str`.

In summary, the `_strdup` function duplicates a given string by determining its length, allocating memory for the duplicated string, copying characters from the input string to the duplicated string, and properly terminating the duplicated string with a null terminator. If successful, it returns a pointer to the duplicated string; otherwise, it returns `NULL` to indicate an error, such as when memory allocation fails or when a `NULL` input string is provided.

**Comment by ahmed149** [comment]
#include <stdio.h>
#include <stdlib.h>
#include <string.h> // Include string.h for strlen function.

/**
 * my_strdup - Duplicate a string
 * @str: The string to duplicate
 * 
 * Return: A pointer to the duplicated string, or NULL on failure.
 */
char *my_strdup(const char *str)
{
    // Check if the input string is NULL.
    if (str == NULL) {
        return NULL; // Return NULL to indicate an error.
    }

    // Calculate the length of the input string.
    size_t len = strlen(str);

    // Allocate memory for the duplicated string.
    char *duplicate = (char *)malloc((len + 1) * sizeof(char));

    // Check if memory allocation was successful.
    if (duplicate == NULL) {
        return NULL; // Return NULL to indicate an error.
    }

    // Copy the characters from the input string to the duplicated string.
    strcpy(duplicate, str);

    return duplicate; // Return a pointer to the duplicated string.
}

int main(void)
{
    const char *original = "Hello, World!";
    
    // Duplicate the string.
    char *duplicate = my_strdup(original);

    if (duplicate != NULL) {
        printf("Original: %s\n", original);
        printf("Duplicated: %s\n", duplicate);

        // Free the allocated memory.
        free(duplicate);
    } else {
        printf("Memory allocation or duplication failed.\n");
    }

    return 0;
}
